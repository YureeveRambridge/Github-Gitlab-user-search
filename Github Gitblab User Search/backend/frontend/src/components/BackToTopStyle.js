import styled from "styled-components";

// Css for the Back to top icon
export const Button = styled.div`
  position: center;
  width: 100%;
  left: 50%;
  bottom: 40px;
  height: 20px;
  font-size: 3rem;
  z-index: 1;
  cursor: pointer;
  color: blue;
`;
