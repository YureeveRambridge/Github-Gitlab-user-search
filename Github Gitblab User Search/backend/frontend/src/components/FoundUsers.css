 /* found users style*/

 /* Positioning of cards */
.resultsDiv{
    display: inline-block;
    margin: 0 35px;
    padding: 3%;
}

/* background*/
.FoundUsersCol{
    background-color: darksalmon;
    padding: 1%;
  }