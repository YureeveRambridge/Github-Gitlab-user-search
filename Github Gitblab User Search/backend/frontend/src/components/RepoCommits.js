import React from "react";
import PropTypes from "prop-types";
import ListGroup from "react-bootstrap/ListGroup";
import dateFormat from 'dateformat';

//Component takes the users commits and displays them.
export function RepoCommits(props) {

  //"commitData" is mapped and displayed.
  const userCommits = props.commitData.map((commit) => {
    return (
      <ListGroup.Item className="listTable">

        {/* Ordered List layout */}
        <li key={commit.id}>
          
          {/* User's message for the commit is displayed */}
          {commit.message} 
          <br></br>

          {/* Date and time for commit is displayed and "dateFormat" function
          converts api date to a readable and time */}
          ({dateFormat(commit.committed_date, "mmmm dS yyyy, hh:mm tt")})
        </li>
      </ListGroup.Item>
    );
  });

  // List of user commits are returned
  return <ol>{userCommits}</ol>;
}

// The required prop is set.
RepoCommits.propTypes = {
  commitData: PropTypes.array.isRequired,
};
