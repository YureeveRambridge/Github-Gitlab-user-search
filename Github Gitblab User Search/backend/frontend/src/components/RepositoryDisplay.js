import React from "react";
import PropTypes from "prop-types";
import Button from 'react-bootstrap/Button';
import ListGroup from 'react-bootstrap/ListGroup';

//Component takes the repository data and displays the list of repositories on the "UserProfile" component.

export function RepositoryDisplay(props) {
  //Extract the list of repos from the given props
  const apiData = props.apiData;

  //Event handler for when the view details button is clicked and runs the "repositoryData" function.
  const eventHandler = (e) => {
    //The vcsProvider and username from the userDetails data
    props.repositoryData(props.userDetails.vcsProvider, props.userDetails.login, e.target.value);
  };

  //User repositories are mapped to be displayed and assigned a view button.
  const repositoryList = apiData.map((repository) => {
    return (
      <ListGroup.Item className="listTable">
       {/* Map Key */}
      <li key={repository.id}>
        <div className="repositoryDisplayList">
          {/* Repository name */}
          <span><b>{repository.name}</b></span>

          {/* View Details Button */}
          <Button
            //When button is clicked the repository name and repository id is passed into repositoryData function.
            // This will bring up the repository's details
            value={[repository.name, repository.id]}
            onClick={eventHandler}
            variant="primary"
          >
            View details
          </Button>
        </div>
      </li>
      </ListGroup.Item>
    );
  });

  //Users repositories and view button are displayed. 
  return <ol className="repositoryList">{repositoryList}</ol>;
}

// The required props and function are set.
RepositoryDisplay.propTypes = {
  repository: PropTypes.array.isRequired,
  userDetails: PropTypes.object.isRequired,
  displayRepo: PropTypes.func.isRequired,
};
