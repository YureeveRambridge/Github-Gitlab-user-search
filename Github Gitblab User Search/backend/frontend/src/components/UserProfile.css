/* Link colour */
a {
    color: green;
  }
  a:hover {
     color: #00F
}

/* Style for list of repos */
.latestRepos{
    padding: 3%;
}