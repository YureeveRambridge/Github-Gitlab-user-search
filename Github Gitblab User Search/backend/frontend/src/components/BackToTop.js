import React, { useState } from "react";
import { FaArrowCircleUp } from "react-icons/fa";
import { Button } from "./BackToTopStyle";

// Function display an icon that when clicked will allow user to scroll back to the top of the page.
const BackToTop = () => {

	// Boolean variable for up arrow button is initilized
  const [upArrowIcon, setUpArrowIcon] = useState(false);

  const scrollerIcon = () => {

	// Gets how much the page has scrolled up
    const scrolledUpAmount = document.documentElement.scrollTop;

	// If the "scrolledUpAmount" is greater than 300 the "upArrowIcon"
	//  variable is set to true and the button is shown.
    if (scrolledUpAmount > 300) {
      setUpArrowIcon(true);
    } 
	// If the "scrolledUpAmount" is less than 300 the "upArrowIcon" 
	//  variable is set to false and the button is not shown.
	else if (scrolledUpAmount <= 300) {
      setUpArrowIcon(false);
    }
  };

  //Function auto scrolls to the top of the page
  const moveUp = () => {
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };
  window.addEventListener("scroll", scrollerIcon);

  return (
    <Button>
		{/* Button icon called from react module react-icons */}
      <FaArrowCircleUp
        onClick={moveUp}
        style={{ display: upArrowIcon ? "inline" : "none" }}
      />
    </Button>
  );
};

export default BackToTop;

// Reference for this function:
// https://www.geeksforgeeks.org/how-to-create-a-scroll-to-top-button-in-react-js/