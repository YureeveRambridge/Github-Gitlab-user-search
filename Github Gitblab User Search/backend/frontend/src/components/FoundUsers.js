import React from "react";
import PropTypes from "prop-types";
import Button from 'react-bootstrap/Button';
import "./FoundUsers.css";
import Card from 'react-bootstrap/Card';

//Component takes found users from the search results and displays it to the web page.
export function FoundUsers(props) {

  //Variable "prop.foundUsers" is simplified
  const foundUsers = props.foundUsers;

  //Event handler for when the "view user" button is clicked. The "viewUser" function will run from the prop data.
  const eventHandler = (e) => {

    //Function will take the website and the username variables and passes it into the function.
    props.viewUser(e.target.value);
  };

  //All found users that's stored in "foundUsers" are mapped, listed in order and displayed on the webpage.
  const usersFound = foundUsers.map((user) => {
    return (
      <div key={user.id} className="resultsDiv">

        <Card style={{ width: '18rem' }} >

          {/* Displays user profile picture */}
          <Card.Img variant="top" src={user.avatar} />

          <Card.Body className="userCard">

          {/* Shows if the username was found on github or gitlab */}
            <Card.Title>{user.vcsProvider}</Card.Title>

            {/* Displays the username */}
            <Card.Text >
              {user.login}
            </Card.Text>

            {/* View button so that the searched user's github/gitlab repositories can be viewed. 
                There are two values that need to be passed to the "viewUser" function. 
                They are passed in an array and will be split in the function it's passed in.
            */}
            <Button 
              value={[user.vcsProvider, user.login]}
              onClick={eventHandler}
              className="viewButton"
              variant="primary"
            >View User
            </Button>
            
          </Card.Body>
        </Card>
      </div>
    );
  });

  //Order list of users are returned
  return <ul>{usersFound}</ul>;
}

//Required props are defined. 
FoundUsers.propTypes = {
  foundUsers: PropTypes.array.isRequired,
  viewUser: PropTypes.func.isRequired,
};
