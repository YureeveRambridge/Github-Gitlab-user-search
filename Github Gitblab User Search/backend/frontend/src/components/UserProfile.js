import React, {useState, useEffect} from "react";
import ListGroup from "react-bootstrap/ListGroup";
import { getJsonData } from "../Get Json Data Function/getJsonData.js";
import { RepositoryDisplay } from "./RepositoryDisplay";
import PropTypes from "prop-types";
import "./UserProfile.css";

//The Component takes the backend api data and displays the details of the users profile from github or gitlab.
// The users profile picture, linked name to his profile on the site are displayed and "RepositoryDisplay" component 
// is called to display the last five repositories of user.

export function UserProfile(props) {

  //Api data variable is initialize
  const [apiData, setApiData] = useState(null);

  //After rendering, the function will run and fetch the data from the sever
  useEffect(() => {

    //Nested async function for synchronous useEffect 
    const getApiData = async () => {

      //url to get the data from the sever. 
      const url = `/api/repositories?vcsProvider=${props.userDetails.vcsProvider}&username=${props.userDetails.login}`;

      // Fetches the data
      const fetchedData = await getJsonData(url);

      //Api data is set
      setApiData(fetchedData);
    };

    //"getApiData" function is called
    getApiData();

    //Effect runs every time the user info changes
  }, [props.userDetails]);

  return (
      <div className="latestRepos">
        <img src={props.userDetails.avatar} alt="userprofilepicture" className="avatar"></img>
        <p>

          {/* Link to user profile on github or gitlab */}
          <a href={props.userDetails.html_url}>
              <h2>{props.userDetails.login}</h2>
          </a>
        </p>

        {/* Heading for list of repositories */}
        <h3>Latest repositories:</h3>

        {/*Loading Display when api request takes long to load */}
        {!apiData ? (
          <img src="https://imgs.search.brave.com/UFRKZlEuTgIy4yfQ2eR6ZG0VL5E7cXsOx3w6ImUVKso/rs:fit:1200:1200:1/g:ce/aHR0cHM6Ly9zdGF0/aWMudmVjdGVlenku/Y29tL3N5c3RlbS9y/ZXNvdXJjZXMvcHJl/dmlld3MvMDAxLzgy/Ni8yNDgvbGFyZ2Vf/MngvcHJvZ3Jlc3Mt/bG9hZGluZy1iYXIt/YnVmZmVyaW5nLWRv/d25sb2FkLXVwbG9h/ZC1hbmQtbG9hZGlu/Zy1pY29uLXZlY3Rv/ci5qcGc" alt="loadingIcon" className="loadingIcon"></img>
        ) : 
        
        /* If the user has no repositories the following is displayed */
        apiData.length === 0 ? (
          <p>No repositories found for user</p>
        ) : (
          
          /* Else the last five users repos are shown by call the component "RepositoryDisplay" */
          <ListGroup>

            <RepositoryDisplay
              userDetails={props.userDetails}
              apiData={apiData}
              repositoryData={props.repositoryData}
            ></RepositoryDisplay>
          </ListGroup>
        )}
      </div>
  );
}

// The required prop and function are set.
UserProfile.propTypes = {
  userDetails: PropTypes.object.isRequired,
  repositoryData: PropTypes.func.isRequired,
};
