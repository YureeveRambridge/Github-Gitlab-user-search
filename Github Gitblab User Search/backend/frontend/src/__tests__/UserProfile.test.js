import React from "react";
import { UserProfile } from "../components/UserProfile";
import renderer from "react-test-renderer";
const fetch = require("node-fetch");

//Test to ensure component renders as wanted
test("User Profile is shown", () => {
  const userDetails = {
    id: 1460315,
    name: "happy",
    login: "happy",
    avatar: "https://avatars.githubusercontent.com/u/1460315?v=4",
    url: "https://api.github.com/users/happy",
    vcsProvider: "Github"
  };
  const snap = renderer
    .create(
      <UserProfile
        userDetails={userDetails}
        repositoryData={function repositoryData() {}}
      />
    )
    .toJSON();
  expect(snap).toMatchSnapshot();
});

//Test get the user details:
test("repositoryData fetches data correctly", () => {
  return fetch("http://localhost:3001/api/viewuser?vcsProvider=Github&username=happy")
    .then((data) => data.json())
    .then((data) => {
      expect(data).toEqual({
        id: 1460315,
        name: "happy",
        login: "happy",
        avatar: "https://avatars.githubusercontent.com/u/1460315?v=4",
        url: "https://api.github.com/users/happy",
        html_url: "https://github.com/happy",
        vcsProvider: "Github"
      });
    });
});
