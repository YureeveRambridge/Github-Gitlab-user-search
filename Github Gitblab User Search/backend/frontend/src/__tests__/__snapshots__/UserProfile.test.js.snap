// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Profile is shown 1`] = `
<div
  className="latestRepos"
>
  <img
    alt="userprofilepicture"
    className="avatar"
    src="https://avatars.githubusercontent.com/u/1460315?v=4"
  />
  <p>
    <a>
      <h2>
        happy
      </h2>
    </a>
  </p>
  <h3>
    Latest repositories:
  </h3>
  <img
    alt="loadingIcon"
    className="loadingIcon"
    src="https://imgs.search.brave.com/UFRKZlEuTgIy4yfQ2eR6ZG0VL5E7cXsOx3w6ImUVKso/rs:fit:1200:1200:1/g:ce/aHR0cHM6Ly9zdGF0/aWMudmVjdGVlenku/Y29tL3N5c3RlbS9y/ZXNvdXJjZXMvcHJl/dmlld3MvMDAxLzgy/Ni8yNDgvbGFyZ2Vf/MngvcHJvZ3Jlc3Mt/bG9hZGluZy1iYXIt/YnVmZmVyaW5nLWRv/d25sb2FkLXVwbG9h/ZC1hbmQtbG9hZGlu/Zy1pY29uLXZlY3Rv/ci5qcGc"
  />
</div>
`;
